File: src/runtime/reaction-economy.ts
// Reaction Economy - the per-round reaction ceiling.
//
// v2.3.0. A combatant gets exactly ONE reaction per round (the rule that kills
// infinite-reaction loops). This is the deterministic, storage-free core: a
// ReactionLedger tracks which combatants have spent their reaction in the
// CURRENT round; advancing the round refreshes everyone.
//
// Robust by construction: each spend is tagged with the round it happened in,
// and canReact compares against the ledger's current round - so even if a
// round-reset is ever missed, a stale spend from a PRIOR round is inert (it can
// never block the current round). advanceReactionRound just increments the
// counter; pruneSpentBefore is an optional memory-bound for very long fights.
//
// Pure + deterministic: no RNG, no wall-clock; an insertion-ordered Map snapshot
// replays identically. The engine owns the ceiling so an AI narrator can never
// grant a combatant a second reaction by describing one.
//
// Code style: var-only, no arrow functions.

// The hard ceiling: reactions available to a combatant at the start of a round.
export const REACTIONS_PER_ROUND = 1;

// round is 1-based (round 1 is the first round of an encounter). spentInRound
// maps an entity id -> the round number in which it last spent its reaction.
export interface ReactionLedger {
  round: number;
  spentInRound: Map<string, number>;
}

// Fresh ledger at the given round (default 1).
export function createReactionLedger(round?: number): ReactionLedger {
  var r = typeof round === 'number' && round > 0 ? Math.floor(round) : 1;
  return { round: r, spentInRound: new Map() };
}

// True iff `entityId` still has its reaction available THIS round.
export function canReact(ledger: ReactionLedger, entityId: string): boolean {
  if (!entityId) return false;
  return ledger.spentInRound.get(entityId) !== ledger.round;
}

// Reactions left for `entityId` this round (0 or REACTIONS_PER_ROUND - the
// ceiling is one, so this is a 0/1 gauge for HUDs).
export function reactionsRemaining(ledger: ReactionLedger, entityId: string): number {
  return canReact(ledger, entityId) ? REACTIONS_PER_ROUND : 0;
}

// Attempt to spend `entityId`'s reaction. Returns true if spent, false if it was
// already spent this round (the ceiling refusing a second reaction) or the id is
// empty. Idempotent within a round: a refused spend changes nothing.
export function spendReaction(ledger: ReactionLedger, entityId: string): boolean {
  if (!entityId) return false;
  if (ledger.spentInRound.get(entityId) === ledger.round) return false;
  ledger.spentInRound.set(entityId, ledger.round);
  return true;
}

// Advance to the next round - every combatant's reaction refreshes (their prior
// spend is now tagged with a stale round). Returns the new round number.
export function advanceReactionRound(ledger: ReactionLedger): number {
  ledger.round += 1;
  return ledger.round;
}

// Force the ledger's round (replay / authoritative sync). A non-positive or
// non-numeric value is ignored.
export function setReactionRound(ledger: ReactionLedger, round: number): void {
  if (typeof round === 'number' && round > 0) ledger.round = Math.floor(round);
}

// Drop spend-records older than the current round (optional memory bound for
// very long encounters). Returns the number of stale records removed. Behavior
// is unchanged either way - stale records are already inert via the round tag.
export function pruneStaleSpends(ledger: ReactionLedger): number {
  var stale: string[] = [];
  for (var entry of ledger.spentInRound) {
    if (entry[1] < ledger.round) stale.push(entry[0]);
  }
  for (var s of stale) ledger.spentInRound.delete(s);
  return stale.length;
}

// Clear every spend-record (encounter end / reset).
export function clearReactions(ledger: ReactionLedger): void {
  ledger.spentInRound.clear();
}

// Deterministic, insertion-ordered snapshot for serialization / replay.
export function reactionLedgerSnapshot(
  ledger: ReactionLedger,
): { round: number; spent: Array<{ entityId: string; round: number }> } {
  var spent: Array<{ entityId: string; round: number }> = [];
  for (var entry of ledger.spentInRound) {
    spent.push({ entityId: entry[0], round: entry[1] });
  }
  return { round: ledger.round, spent: spent };
}

// Resource key for the world's resource registry.
export const RESOURCE_REACTION_ECONOMY = 'reactionEconomy';

File: python/loom_engine/reaction_economy.py
"""loom_engine.reaction_economy - the per-round reaction ceiling.

Byte-parity hand-port of the TypeScript runtime/reaction-economy.ts: exactly ONE
reaction per combatant per round, each spend round-tagged so a stale prior-round
record is inert. Pure + deterministic (no RNG, no wall-clock; ordered dict
matches the TS Map). See ../test_vectors/ for the cross-language gate.
"""

from __future__ import annotations

from typing import Dict, List

REACTIONS_PER_ROUND = 1


class ReactionLedger:
    """round is 1-based. spent_in_round maps entity_id -> the round it last spent
    its reaction (ordered-insertion dict, matching the TS Map)."""

    def __init__(self, round_no: int = 1) -> None:
        self.round = int(round_no) if isinstance(round_no, int) and round_no > 0 else 1
        self.spent_in_round: Dict[str, int] = {}


def create_reaction_ledger(round_no: int = 1) -> ReactionLedger:
    return ReactionLedger(round_no)


def can_react(ledger: ReactionLedger, entity_id: str) -> bool:
    """True iff entity_id still has its reaction available THIS round."""
    if not entity_id:
        return False
    return ledger.spent_in_round.get(entity_id) != ledger.round


def reactions_remaining(ledger: ReactionLedger, entity_id: str) -> int:
    return REACTIONS_PER_ROUND if can_react(ledger, entity_id) else 0


def spend_reaction(ledger: ReactionLedger, entity_id: str) -> bool:
    """Spend the reaction. True if spent, False if already spent this round (the
    ceiling refusing a second) or the id is empty."""
    if not entity_id:
        return False
    if ledger.spent_in_round.get(entity_id) == ledger.round:
        return False
    ledger.spent_in_round[entity_id] = ledger.round
    return True


def advance_reaction_round(ledger: ReactionLedger) -> int:
    """Advance to the next round - everyone's reaction refreshes (prior records
    become stale). Returns the new round number."""
    ledger.round += 1
    return ledger.round


def set_reaction_round(ledger: ReactionLedger, round_no: int) -> None:
    if isinstance(round_no, int) and round_no > 0:
        ledger.round = round_no


def prune_stale_spends(ledger: ReactionLedger) -> int:
    """Drop spend-records older than the current round (memory bound). Behavior
    unchanged - stale records are already inert. Returns the count removed."""
    stale = [eid for eid, r in ledger.spent_in_round.items() if r < ledger.round]
    for eid in stale:
        del ledger.spent_in_round[eid]
    return len(stale)


def clear_reactions(ledger: ReactionLedger) -> None:
    ledger.spent_in_round.clear()


def reaction_ledger_snapshot(ledger: ReactionLedger) -> dict:
    """Deterministic, insertion-ordered snapshot for serialization / replay."""
    return {
        "round": ledger.round,
        "spent": [{"entity_id": eid, "round": r}
                  for eid, r in ledger.spent_in_round.items()],
    }


RESOURCE_REACTION_ECONOMY = "reactionEconomy"
